certora/specs/HealthStatusInvariant.spec

```diff
@@ -0,0 +1,179 @@
+*  - In controlCollateral the from address is used to set the onBehalfOfAccount
+*  and some authentication is done
+* - After this, callWithContextInternal invokes the transfer function 
+* on the collateral address
+* - Collaterals in the EVK must all be Token.sol and token's transfer
+* implementation calls initOperation which enqueues an account status
+* check on the EVC for the onBehalfOfAccount it also gets from EVC.
+* Because onBehalfOfAccount was set to the from address in callWithContextInternal this status check is for the from account
+* To emulate this, we:
+* - explicitly call EToken.transferFrom using the expected addresses
+* - enqueue a status check on the evc for the "from" address
+*/
+function CVLEnforceCollateralTransfer(env e, address collateral, uint256 amount, address from, address receiver) {
+    evc.requireAccountStatusCheck(e, from);
+    if (collateral == ETokenA) {
+        ETokenA.transferFromInternalHarnessed(e, from, receiver, amount);
+    } else if (collateral == ETokenB) {
+        ETokenB.transferFromInternalHarnessed(e, from, receiver, amount);
+    }
+}
+
+// Assuming the prices stay the same, a healthy account can never become 
+// unhealthy. Here, our assumption that the prices do not change is implicit
+// in the fact that the summary for GetQuote is an uninterpreted function --
+// the prover will model it as a function so it will always return the same
+// value when given the same arguments.
+rule accountsStayHealthy_strategy (method f) filtered { f -> 
+    // Literal selectors are used to avoid compilation errors when
+    // only some of the modules are in the verification scene
+    // sig:GovernanceModule.clearLTV(address).selector
+    f.selector != 0x8255d029 &&
+    // sig:GovernanceModule.setLTV(address,uint16,uint16,uint32).selector
+    f.selector != 0x4bca3d5b &&
+    // sig:InitializeModule.initialize(address).selector
+    f.selector != 0xc4d66de8 &&
+    // sig:TokenHarnes.transferFromInternalHarnessed (this is a harness method only)
+    f.selector != 0xd3110e86
+}{
+    env e;
+    calldataarg args;
+    address account;
+    address[] collaterals = evc.getCollaterals(e, account);
+    require collaterals.length <= 2; // loop bound
+    require oracleAddress != 0; 
+    // not sure the following 4 are really needed
+    require account != erc20;
+    require account != oracleAddress;
+    require account != evc;
+    require account != unitOfAccount;
+
+    require LTVConfigAssumptions(e, getLTVConfig(e, ETokenA));
+    require LTVConfigAssumptions(e, getLTVConfig(e, ETokenB));
+    if (collaterals.length > 0) {
+        require collaterals[0] == ETokenA || collaterals[0] == ETokenB;
+    }
+    if (collaterals.length > 1) {
+        require collaterals[1] == ETokenA || collaterals[1] == ETokenB;
+    }
+
+    bool healthyBefore = checkLiquidityReturning(e, account, collaterals);
+    f(e, args);
+    // The only way to call a vault funciton is through EVC's call, batch, 
```

Comment:
This assert is what saves you in practice, yes. To prove this property we are essentially assuming this assertion is there.

Reply:
I must be missing something, but I am wondering how come this rule is passing in this form though. The assertion mentioned only enforces that `msg.sender == evc`, but from the prover POV, it doesn't enforce that the EVC's execution context is properly set (I mean the checks in progress flag). When the vault requires the account status check using the `requireAccountAndVaultStatusCheck` function (which is dispatched) and that flag is not set, the check is performed immediately. It means that we perform it before the actual operation (as it is requested by the `initOperation` function in the vault, not at the end). 

Because of that, I am wondering how is it possible that we're not getting any violation here. If the account status check is performed before the operation and not after, it is possible that we leave the account unhealthy.

Reply:
You are right. For this rule I am making _assumptions_ about what the EVC does in call/batch:
* the EVC is the only way to invoke the vault (either directly or indirectly)
* the status checks are deferred (this is in the CVL summary `CVLAreChecksInProgress`)
* the EVC always executes any deferred status checks at the end of call/batch
These things can all be covered by separate properties about the EVC.
The last one we did also prove separately already. The first two should be simpler ones but we did not prove these. I did not already prove all of these because these all seem relatively clearer/simpler and I recall in the project Euler wanted to make sure the verification time was spent on more difficult rules that are less clear to check from audit.

Reply:
>the status checks are deferred (this is in the CVL summary `CVLAreChecksInProgress`)

I don't think that's correct. in the EVC's execution context, checks deferred are indicated by a different flag than checks in progress. `CVLAreChecksInProgress` is needed in order for `checkAccountStatus`/`checkVaultStatus` functions on the vault not to revert. however, we have no guarantee now that the checks are really deferred because there's no summary for `EVC.areChecksDeferred` function at the moment. this is why I am wondering why this rule passes and we're not getting any counterexample

Reply:
Hi @kasperpawlowski. Thanks so much for pointing this out and for your thorough review. You are absolutely right, and this helped us identify a mistake in the spec. It should need and include an assumption that checks are deferred here, but there was another issue with the spec. This was subtle and we have been investigating it. Fortunately, we think we have it fixed now, and are rerunning the fixed rules to confirm that our fix is correct.

It turns out that due to a mistake in our summaries, with this setup we could spuriously prove EVC.areChecksDeferred() returns true in all cases: https://vaas-stg.certora.com/output/65266/0f25180bc5304f6c9c1f3a1f3e53a195/?anonymousKey=6b7327f914bb55c04494f19c57f409a2045bc1c2. The reason for this is that with the env variables in our CVLCheckAccountStatusInternal summary as they were written when you reviewed this we actually caused checkAccountStatus@withRevert to always revert:


```
function CVLCheckAccountStatusInternalBool(env e, address account) returns bool {
    address[] collaterals = evc.getCollaterals(e, account);
    checkAccountStatus@withrevert(e, account, collaterals); // the problem is with this env variable more details below.
    return !lastReverted;
}

// This summary is applied to EthereumVaultConnector.checkAccountStatusInternal
function CVLCheckAccountStatusInternal(env e, address account) returns (bool, bytes) {
    return (CVLCheckAccountStatusInternalBool(e, account), 
        checkAccountMagicValueMemory(e));
}
```

The issue with this is that here we are passing the env e that was used during the the vault’s call to EVC’s checkAccountStatus.Because env variables have the msg.sender in them, this forces the msg.sender to be the the same for both: 1) the call from the contract under verification (a vault module) to EVC.requireAccountStatusCheck (which is reached as part of Base.init’s call to EVCRequireStatusChecks) and 2) the call back from EVC into the vault (the call from EVC.requireAccountStatusCheckInternal into Riskmanager.checkAccountStatus). As a result, msg.sender will be the contract under verification during RiskManager.checkAccountStatus rather than evc and when we reach this revert check as a part of the checkAccountStatus call it will always revert

```
    function checkAccountStatus(address account, address[] calldata collaterals)
        public
        virtual
        reentrantOK
        onlyEVCChecks /// this is the important part
        returns (bytes4 magicValue)
    {
    }

    modifier onlyEVCChecks() {
        if (msg.sender != address(evc) || !evc.areChecksInProgress()) { 
            revert E_CheckUnauthorized(); // we were always reverting here during just the EVC checkAccount/Vault summaries
        }

        _;
    }
```

There is a similar problem with both the summary for checkAccountStatusInternal and checkVaultStatusInternal.

We confirmed this was the issue by declaring fresh environment variables for these summaries and continuing to omit the requirement that evc.areChecksDeferred().In this case there is a counterexample as you would expect:  https://vaas-stg.certora.com/output/65266/7663d91b72c04214898586e968adee89/?anonymousKey=a9b11593bd253324353519853241b692b2f82a24 Note also that in this failing case, evc.areChecksDeferred is returning false indicating this is the reason for the counterexample.

Here is the code that declares a fresh environment variable in the summary. Our candidate fix is to make this change and also add the assumption that evc.areChecksDeferred:

```
function CVLCheckAccountStatusInternal(address account) returns (bool, bytes) {
    // We need a new env for the first function.
    // Since the vault calls the EVC, otherwise msg.sender
    // would become the vault unless we declare a fresh environment.
    env eEVC;
    return (CVLCheckAccountStatusInternalBool(eEVC, account), 
        checkAccountMagicValueMemory());
}
```

We are currently rerunning all the holy grail runs to make sure that it still passes with this fix. I do want to assure you though that despite this problem and investigation, I did not suspect there was a critical issue with the main intent of the rule – during the development of the rule it failed unless we called evc.checkStatusAll to make the EVC’s call to this function (as a part of restoring the execution context) at the end of call/batch. This is a good sign because it means the rule would catch any issues related to not enqueueing the right deferred account checks. The issue was just that we subtly forced the deferred status checks flag to always be true.

That said, we take the quality of our specifications and verification results very seriously and:
We are rerunning all the holy grail runs to confirm it still passes after this fix
We will thoroughly re-review all of our specifications here
We will write the additional rules you requested and potentially a few others related to assumptions made about the EVC.

Thanks again for your own very detailed and thorough review of the specifications.